### ISSUE TITLE
Emscripten version

### ISSUE THREAD

**Issue by Antom91** [FWGS/xash3d-fwgs#1548]
Есть ли какие то примеры как собрать версию Emscripten свежего кода ? либо кто может помочь это сделать , возможно за отдельную плату.

Либо кто то уже пытался? какой статус этих попыток и т.д.

---
Are there any examples of how to build a version of Emscripten with fresh code? or who can help do this, perhaps for a fee.

Or has someone already tried it? what is the status of these attempts, etc.

**Comment by a1batross** [comment]
Насколько я знаю, нынешнюю ветку никто под Emscripten никогда не собирал.

В любом случае, производительности JS/WASM в браузерах редко хватает на полный движок, плюс есть определенные ограничения, вроде отсутствия UDP необходимого мультиплееру (есть только костыль с WebSockets, WebRTC для сетевой игры пока никто не пытался применить) или мелочи вроде необходимости поддержки 48000 Hz в микшере, которой у нас нет.

**Comment by mittorn** [comment]
Не знаю, как насчёт wasm, но произврдительности asmjs вполне хватало ещё в 2017, причём на браузерах нескольколетней давности. Логично предположить что с wasm дела должны обстоять как минимум не хуже. Из реальных нерешаемых проблем - сейчас многие недобраузеры заразились хромовской чумой с закрытием вкладки по ctrl+w, так что для нормальной игры потребуется патчить браузер.
Рендер упирается в произврдительность скиннинга моделей в мультиплеере, при этом в сингле особых проблем нет.
В целом, в движке всё есть для поддержки emscripten, мало того, багов с выравниванием меньше по сравнению со старым движком. Так что, нужно видимо всего лишь собрать всё это во что-то рабочее и переписать платформенный код загрузки библиотек

**Comment by Antom91** [comment]
Было бы отлично если бы кто то попробовал это сделать. Много ли там работы ? человеку, который не работал с этим я полагаю будет сложно :-( 

А по поводу CTRL+W то, возможно можно эту штуку игнорировать на уровне Javascript ? есть же play-cs , как то решили это.

Вижу как работает старый emscripten который вроде как [mittorn](https://github.com/mittorn) делал, работает супер + мультиплеер работает. Туда бы поддержку сstrike добавить, и было бы супер!

**Comment by nekonomicon** [comment]
> есть же play-cs , как то решили это.

Так там просто старые движок и cs16client.
Нынешняя администрация - паразиты и кидалы, вряд ли они там что-то кардинально меняли.

**Comment by mittorn** [comment]
А точно ли это как-то решили в play-cs? Может за эти годы в браузерах что-то поменяли?

**Comment by mittorn** [comment]
Насчёт cstrike: там рендер тормозил из-за сложных мрделей. Чтобы было играбельно, скорее всего придётся упроститт модели, либо писать кастомный рендер со скиннингом на gpu как в xashxt

**Comment by Antom91** [comment]
> > есть же play-cs , как то решили это.
> 
> Так там просто старые движок и cs16client. Нынешняя администрация - паразиты и кидалы, вряд ли они там что-то кардинально меняли.

да, движок старый , и я пробовал дебажить, никто ничего там не менял, допилили просто к тому что уже был и всё. Но работает )

> А точно ли это как-то решили в play-cs? Может за эти годы в браузерах что-то поменяли?

сказать точно не могу, как только нажимаешь ctrl+w - сразу игра переходит в фулскрин, но работает, обратил внимание

Да и кастомные движки разные пробовал на гитхабе , в некоторых тоже эта комбинация не закрывает складку 



> Насчёт cstrike: там рендер тормозил из-за сложных мрделей. Чтобы было играбельно, скорее всего придётся упроститт модели, либо писать кастомный рендер со скиннингом на gpu как в xashxt

да хоть как то что бы завелось для начала, было бы шикарно, жаль что у меня нет опыта, а у остальных желания :(  Хороший бы проект получился , удобный, портативный

**Comment by Antom91** [comment]
Все ещё актуально, может есть какие то новости по поводу ?

**Comment by Ser-Gen** [comment]
This browser API can help with Ctrl+W

https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/lock
